testing/integration/kubernetes_agent_standalone_test.go

```diff
@@ -937,43 +861,260 @@ func generateESAPIKey(esClient *elasticsearch.Client, keyName string) (string, e
+	timeoutCtx, timeoutCancel := context.WithTimeout(ctx, opts.waitTimeout)
+	defer timeoutCancel()
 	for _, obj := range objects {
-		if timeoutCtx.Err() != nil {
-			break
-		}
-
-		for i := 0; i < 10; i++ {
+		for {
 			if timeoutCtx.Err() != nil {
-				break
+				return errors.New("timeout waiting for k8s objects to be deleted")
 			}
 
 			err := client.Resources().Get(timeoutCtx, obj.GetName(), obj.GetNamespace(), obj)
 			if err != nil {
+				// object has been deleted
 				break
 			}
-			time.Sleep(500 * time.Millisecond)
+
+			time.Sleep(100 * time.Millisecond)
 		}
 	}
 
-	if timeoutCtx.Err() != nil {
-		t.Log("Timeout waiting for k8s objects to be deleted")
-	}
+	return nil
 }
 
+// int64Ptr returns a pointer to the given int64
 func int64Ptr(val int64) *int64 {
 	valPtr := val
 	return &valPtr
 }
+
+// k8sCreateOpts contains options for k8sCreateObjects
+type k8sCreateOpts struct {
+	// namespace is the namespace to create the objects in
+	namespace string
+	// wait specifies whether to wait for the objects to be ready
+	wait bool
+	// waitTimeout is the timeout for waiting for the objects to be ready if wait is true
+	waitTimeout time.Duration
+}
+
+// k8sCreateObjects creates k8s objects and waits for them to be ready if specified in opts.
+// Note that if opts.namespace is not empty, all objects will be created and updated to reference
+// the given namespace.
+func k8sCreateObjects(ctx context.Context, client klient.Client, opts k8sCreateOpts, objects ...k8s.Object) error {
+	// Create the objects
+	for _, obj := range objects {
+		if opts.namespace != "" {
+			// update the namespace
+			obj.SetNamespace(opts.namespace)
+
+			// special case for ClusterRoleBinding and RoleBinding
+			// update the subjects to reference the given namespace
+			switch objWithType := obj.(type) {
+			case *rbacv1.ClusterRoleBinding:
+				for idx := range objWithType.Subjects {
+					objWithType.Subjects[idx].Namespace = opts.namespace
+				}
+				continue
+			case *rbacv1.RoleBinding:
+				for idx := range objWithType.Subjects {
+					objWithType.Subjects[idx].Namespace = opts.namespace
+				}
+				continue
+			}
+		}
+		if err := client.Resources().Create(ctx, obj); err != nil {
+			return fmt.Errorf("failed to create object %s: %w", obj.GetName(), err)
+		}
+	}
+
+	if !opts.wait {
+		// no need to wait
+		return nil
+	}
+
+	if opts.waitTimeout == 0 {
+		// default to 120 seconds
+		opts.waitTimeout = 120 * time.Second
+	}
+
+	return k8sWaitForReady(ctx, client, opts.waitTimeout, objects...)
+}
+
+// k8sWaitForReady waits for the given k8s objects to be ready
+func k8sWaitForReady(ctx context.Context, client klient.Client, waitDuration time.Duration, objects ...k8s.Object) error {
+	// use ready checker from helm kube
+	clientSet, err := kubernetes.NewForConfig(client.RESTConfig())
+	if err != nil {
+		return fmt.Errorf("error creating clientset: %w", err)
+	}
+	readyChecker := helmKube.NewReadyChecker(clientSet, func(s string, i ...interface{}) {})
```

Comment:
Out of curiosity, what does this actually do for resources which don't have an innate notion of readiness in K8s?

Reply:
here you go https://github.com/helm/helm/blob/main/pkg/kube/ready.go#L204 🙂